federation/pkg/federation-controller/sync/controller.go

```diff
@@ -453,33 +453,70 @@ func syncToClusters(clustersAccessor clustersAccessorFunc, operationsAccessor op
 type clusterObjectAccessorFunc func(clusterName string) (interface{}, bool, error)
 
 // clusterOperations returns the list of operations needed to synchronize the state of the given object to the provided clusters
-func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, obj pkgruntime.Object, accessor clusterObjectAccessorFunc) ([]util.FederatedOperation, error) {
-	key := federatedtypes.ObjectKey(adapter, obj)
-	operations := make([]util.FederatedOperation, 0)
+func clusterOperations(adapter federatedtypes.FederatedTypeAdapter, clusters []*federationapi.Cluster, fedObj pkgruntime.Object, accessor clusterObjectAccessorFunc) ([]util.FederatedOperation, error) {
+	key := federatedtypes.ObjectKey(adapter, fedObj)
+	currentObjs := make(map[string]pkgruntime.Object)
 	for _, cluster := range clusters {
 		clusterObj, found, err := accessor(cluster.Name)
 		if err != nil {
 			wrappedErr := fmt.Errorf("Failed to get %s %q from cluster %q: %v", adapter.Kind(), key, cluster.Name, err)
 			runtime.HandleError(wrappedErr)
 			return nil, wrappedErr
 		}
-		// The data should not be modified.
-		desiredObj := adapter.Copy(obj)
 
-		var operationType util.FederatedOperationType = ""
 		if found {
-			clusterObj := clusterObj.(pkgruntime.Object)
-			if !adapter.Equivalent(desiredObj, clusterObj) {
+			currentObjs[cluster.Name] = clusterObj.(pkgruntime.Object)
+		} else {
+			currentObjs[cluster.Name] = nil
+		}
+	}
+
+	scheduledObjs := make(map[string]pkgruntime.Object)
+	if adapter.IsScheduler() {
+		schedAdapter, ok := adapter.(federatedtypes.SchedulerAdapter)
+		if !ok {
+			return nil, fmt.Errorf("%s adapter does not implement required method", adapter.Kind())
+		}
+		var err error
+		// The adapter implementation can choose to update the federated object status, if need be.
+		// TODO: is this really needed as a separate method ?
+		err = schedAdapter.FedUpdateStatus(fedObj, currentObjs)
+		if err != nil {
+			return nil, fmt.Errorf("Failure in updating federated object status %s %q : %v", adapter.Kind(), key, err)
+		}
+		scheduledObjs, err = schedAdapter.Schedule(fedObj, currentObjs)
+		if err != nil {
+			return nil, fmt.Errorf("Failure executing Schedule Hook on %s %q: %v", adapter.Kind(), key, err)
+		}
+	} else {
+		for clusterKey := range currentObjs {
+			// The same object propagated into all clusters.
+			desiredObj := adapter.Copy(fedObj)
```

Comment:
We pass the existing cluster local objects to schedule method(s) together with fedObj.
IMHO, it probably is reasonable expectation on schedule to copy objects into target, where it needs to create/modify them. At this stage its anyways unknown, where which object would go (fedObj to be copied or local obj to be copied).
This copy is only from fed object to the objects which need to percolate to all clusters.